### PR TITLE
Workbox v5 Alpha + Pre-cache Simplification

### PR DESCRIPTION
**Pull request by derekdowling**
Addresses the issues as raised in https://github.com/hanford/next-offline/issues/156. As outlined in the issue, this required upgrading to [Workbox v5 Alpha](https://github.com/GoogleChrome/workbox/releases/tag/v5.0.0-alpha.0) in order to leverage the new pre-cache approach allowing us to leverage their manifest resource prefixing config options. This alleviates the need for the Manifest plugin and cleans up a lot of complexity in general.

Note:
When testing locally with `yarn link`, the paths in the pre-cache manifest of the generated Service Worker file in the `tests` directory are pretty funky. Unlinking causes these files to generate with sane paths as expected.

TODO:
- [x] validate examples/next-export works as expected with the production SW
- [x] validate examples/now1 works as expected with the production SW
- [ ] validate examples/now2 works as expected with the production SW

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,12 +54,10 @@
     "next": ">=7.0.0"
   },
   "dependencies": {
-    "clean-webpack-plugin": "^0.1.19",
```
Comment (derekdowling) [inline package.json]: No longer necessary.

File: package.json [hunk 2]
```diff
@@ -54,12 +54,10 @@
...
     "copy-webpack-plugin": "~4.5.2",
     "fs-extra": "~7.0.0",
-    "glob": "~7.1.3",
```
Comment (derekdowling) [inline package.json]: Unused as far as I can tell, happy to revert this change if it is out of scope.

File: tests/package.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
     "dev": "node server.js",
     "clean": "rm -rf .next static/manifest",
     "build": "next build",
-    "start": "NODE_ENV=production node server.js",
+    "start": "yarn clean && yarn build && NODE_ENV=production node server.js",
```
Comment (derekdowling) [inline tests/package.json]: Hopefully this is okay, again, happy to revert this change, but it made the feedback loop considerably quicker whiel testing.

File: tests/test.js [hunk 4]
```diff
@@ -42,3 +42,28 @@ test('build next app with manifest and service worker', async () => {
   const str = await read(join(cwd, '.next/main.js'), 'utf8');
   expect(str).toEqual(expect.stringContaining('serviceWorker'));
 });
+
+test('pre-caching includes static assets and bundles', async () => {
```
Comment (derekdowling) [inline tests/test.js]: Added a new test case 🎉

File: tests/test.js [hunk 5]
```diff
@@ -42,3 +42,28 @@ test('build next app with manifest and service worker', async () => {
...
+  const str = await read(join(cwd, '.next/service-worker.js'), 'utf8');
+
+  // Check that various bundles are getting entered into pre-cache manifest
+  expect(str).toEqual(expect.stringContaining('/pages/_app.js'));
+  expect(str).toEqual(expect.stringContaining('_next/static/chunks/commons.'));
+
+  // Check that static asset copying via glob pattern is working as expected
+  expect(str).toEqual(expect.stringContaining('_next/static/image.jpg'));
+
+  // TODO: determine whether we should be excluding this file. Not sure if it should be pre-cached
```
Comment (derekdowling) [inline tests/test.js]: @hanford any thoughts? It appears this was previously being pre-cached but I just want to confirm this is desired. Admittedly, I'm just entering the PWA world so I'm not sure if this is a best practice or what.

File: tests/test.js [hunk 6]
```diff
@@ -42,3 +42,28 @@ test('build next app with manifest and service worker', async () => {
...
+          name: 'next-app',
+        },
+      }),
+    ),
+  );
+
+  await nextBuild.default(cwd, nextConf);
+  const str = await read(join(cwd, '.next/service-worker.js'), 'utf8');
+
+  // Check that various bundles are getting entered into pre-cache manifest
```
Comment (derekdowling) [inline tests/test.js]: There are build hashes in the two assertions below that I avoided. I'm sure I could pipe the build ID into this test file, and will if you'd prefer that, but this is a good start.

File: index.js [hunk 7]
```diff
@@ -53,15 +64,10 @@ module.exports = (nextConfig = {}) => ({
     } else if (!options.isServer) {
       // Only run once for the client build.
       config.plugins.push(
-        new CleanWebpackPlugin(['precache-manifest.*.js'], { root: config.output.path, verbose: false }),
```
Comment (derekdowling) [inline index.js]: No longer necessary.

File: index.js [hunk 8]
```diff
@@ -53,15 +64,10 @@ module.exports = (nextConfig = {}) => ({
...
+        // Workbox uses Webpack's asset manifest to generate the SW's pre-cache manifest, so we need
+        // to copy the app's assets into the Webpack context so those are picked up.
+        new CopyWebpackPlugin([{ from: `${join(cwd(), nextAssetPrefixDir)}/**/*` }]),
         generateSw ? new GenerateSW({ ...workboxOpts }) : new InjectManifest({ ...workboxOpts }),
-        new InlineNextPrecacheManifestPlugin({
```
Comment (derekdowling) [inline index.js]: No longer necessary, the new `exclude/modifyURLPrefix` options handle all of this now.

**Comment by derekdowling** [comment]
When I run `yarn build` in the `tests` package I get:
```
➜  tests git:(workbox-v5-precaching) yarn build
yarn run v1.16.0
warning package.json: No license field
$ next build
[13:19:06] Compiling server
[13:19:06] Compiling client
Browserslist: caniuse-lite is outdated. Please run next command `yarn upgrade caniuse-lite browserslist`
[13:19:07] Compiled server in 993ms
'/Users/dowderek/go/src/code.justin.tv/web/next-offline/node_modules/workbox-build/node_modules/workbox-expiration/Plugin.mjs' is imported by ../../../../../../../../private/var/folders/yc/kz52p_t911lgb94vszfftn91skbwm_/T/e6d57a52cbbc64c08f919301ad32cadf/service-worker.js, but could not be resolved – treating it as an external dependency
'/Users/dowderek/go/src/code.justin.tv/web/next-offline/node_modules/workbox-build/node_modules/workbox-strategies/NetworkFirst.mjs' is imported by ../../../../../../../../private/var/folders/yc/kz52p_t911lgb94vszfftn91skbwm_/T/e6d57a52cbbc64c08f919301ad32cadf/service-worker.js, but could not be resolved – treating it as an external dependency
'/Users/dowderek/go/src/code.justin.tv/web/next-offline/node_modules/workbox-build/node_modules/workbox-routing/registerRoute.mjs' is imported by ../../../../../../../../private/var/folders/yc/kz52p_t911lgb94vszfftn91skbwm_/T/e6d57a52cbbc64c08f919301ad32cadf/service-worker.js, but could not be resolved – treating it as an external dependency
'/Users/dowderek/go/src/code.justin.tv/web/next-offline/node_modules/workbox-build/node_modules/workbox-precaching/precacheAndRoute.mjs' is imported by ../../../../../../../../private/var/folders/yc/kz52p_t911lgb94vszfftn91skbwm_/T/e6d57a52cbbc64c08f919301ad32cadf/service-worker.js, but could not be resolved – treating it as an external dependency
[13:19:09] Compiled client in 2s
✨  Done in 5.85s.
```

I still need to dig into this.

File: tests/package.json [hunk 9]
```diff
@@ -13,7 +13,7 @@
     "next": "7.0.2",
     "next-compose-plugins": "^2.1.1",
     "next-manifest": "^1.6.0",
-    "next-offline": "latest",
+    "next-offline": "https://github.com/derekdowling/next-offline#e910c1b",
```
Comment (derekdowling) [inline tests/package.json]: We'll want to revert this before merging I guess. This is where a monorepo would be handy 😆

**Comment by hanford** [comment]
@derekdowling this is looking extremely exciting! Will dive into it tonight.

**Comment by hanford** [comment]
So @derekdowling did some testing and have some questions.

First of all, super excited for this work to land .. looks like workbox v5 is going to have some solid additions and removing the manifest stuff is pretty big win for us.

So I pulled the changes down and the tests were working as expected which is good.

Then I tried bringing your fork into the examples and had some trouble. Noticed the service worker either wasn't registering properly or was 404'ing for both the Now 1.0 and Now 2.0 examples. I was having similar results on my [personal website](https://github.com/hanford/personal-website).

Any ideas?

**Comment by hanford** [comment]
Looks like 5.0 alpha 1 was released last night

https://github.com/GoogleChrome/workbox/releases/tag/v5.0.0-alpha.1

**Comment by derekdowling** [comment]
@hanford already upgrade to alpha.1, I'll dive into the examples today to see what's up.

**Comment by derekdowling** [comment]
@hanford I'm admittedly not very experience with either version 1 or 2 of Now but am continuing to dig in. If you have any example screenshots of the error you are seeing that would be super helpful.

File: examples/now2/next.config.js [hunk 10]
```diff
@@ -5,6 +5,9 @@ const withOffline = moduleExists('next-offline')
 const nextConfig = {
   target: "serverless",
   workboxOpts: {
+    // Trying to set NODE_ENV=production when running yarn dev causes a build-time error so we
```
Comment (derekdowling) [inline examples/now2/next.config.js]: @hanford presumably I should revert this before merging? It would be nice if these examples had some "prod-like" command so that you can test locally but with the real deal SW.

File: tests/package.json [hunk 11]
```diff
@@ -4,16 +4,16 @@
...
     "test": "yarn clean && jest",
     "export": "yarn build && next export"
   },
   "dependencies": {
     "mkdirp": "^0.5.1",
     "next": "7.0.2",
     "next-compose-plugins": "^2.1.1",
-    "next-manifest": "^1.6.0",
-    "next-offline": "latest",
+    "next-manifest": "^2.0.0",
```
Comment (derekdowling) [inline tests/package.json]: This wasn't strictly necessary, but since I was messing with all this stuff, I wanted to make sure nothing was broken here.

**Comment by derekdowling** [comment]
Looks like the current issue is:
```
service-worker.js:1 Uncaught (in promise) DOMException: Failed to execute 'importScripts' on 'WorkerGlobalScope': Failed to import 'http://localhost:3000/workbox-68b0850e.js'. importScripts() of new scripts after service worker installation is not allowed.
    at http://localhost:3000/service-worker.js:1:90
    at async Promise.all (index 0)
```
Looks like we need to tell Workbox to put its assets in `.next/static` presumably.

**Comment by hanford** [comment]
@derekdowling I can do some additional digging and report back tonight or tomorrow.

What are you using to test this, just the test suite?

`Now` isn't doing anything special for `next-offline` which is why I'd imagine it should just work.

But what I'm not seeing the service worker being generated / served at all which is really strange .. again I can get more information and report back.

Thanks again!

File: tests/package.json (hunk 11 above)
Comment (hanford) [reply tests/package.json]: eh, this is probably ok.

If we can actually entirely remove `next-manifest` it might be nice.

I added it mainly to test compatibility w/ `next-compose-plugins`

File: examples/next-export/package.json [hunk 12]
```diff
@@ -3,12 +3,14 @@
...
+    "dev:prodlike": "yarn export && serve -p 3000 ./out"
   },
   "dependencies": {
     "next": "^8.1.0",
-    "next-offline": "latest",
+    "next-offline": "https://github.com/derekdowling/next-offline#0270acd",
     "react": "^16.8.6",
-    "react-dom": "^16.8.6"
+    "react-dom": "^16.8.6",
+    "serve": "^11.0.1"
```
Comment (derekdowling) [inline examples/next-export/package.json]: This sets up a static server allowing us to run the app in export mode in order to test the Service Worker locally 🎉

**Comment by brad-decker** [comment]
oh wow, this exciting. :D :D :D

File: index.js [hunk 13]
```diff
@@ -18,17 +29,23 @@ module.exports = (nextConfig = {}) => ({
...
       generateInDevMode = false,
-      transformManifest = manifest => manifest,
       workboxOpts = {
-        globPatterns: ['static/**/*'],
-        globDirectory: '.',
+        exclude: preCacheManifestBlacklist,
+        modifyURLPrefix: {
+          'static/': `${nextUrlAssetPrefix}/static/`,
+        },
+        manifestTransforms: [
```
Comment (derekdowling) [inline index.js]: This is just a debugging aid, needs to be removed before merging.

**Comment by derekdowling** [comment]
@hanford I think this is ready for review. I was able to validate tests, examples/next-export, and examples/now1 were all working as intended using prod-like builds (`NODE_ENV=production`).

Not sure how to test `now2` locally in prod mode to trigger the service worker logic though because I get an error:
```
yarn run v1.16.0
warning package.json: No license field
$ NODE_ENV=production next
[ wait ]  starting the development server ...
[ info ]  waiting on http://localhost:3000 ...
Error: Cannot start server when target is not server. https://err.sh/zeit/next.js/next-start-serverless
```

Besides that I think this is GTG 👍

File: tests/package.json [hunk 14]
```diff
@@ -4,21 +4,21 @@
...
+    "next-offline": "https://github.com/derekdowling/next-offline#d3230ea",
     "react": "16.6.1",
     "react-dom": "16.6.1",
     "rimraf": "^2.6.2"
   },
   "devDependencies": {
-    "jest": "^23.6.0"
+    "jest": "^24.8.0"
```
Comment (derekdowling) [inline tests/package.json]: I was having some issues that I thought were Jest config related so I bumped to latest. Turns out that wasn't the case. Happy to revert though.

File: index.js [hunk 15]
```diff
@@ -18,17 +29,23 @@ module.exports = (nextConfig = {}) => ({
...
+      // https://developers.google.com/web/ilt/pwa/introduction-to-service-worker#registration_and_scope
       registerSwPrefix = '',
       scope = '/',
-      generateInDevMode = false,
-      transformManifest = manifest => manifest,
       workboxOpts = {
-        globPatterns: ['static/**/*'],
-        globDirectory: '.',
+        exclude: preCacheManifestBlacklist,
+        // TODO: Do we want to bundle Workbox inline? Makes the SW + dependencies slightly less cache-able, but it
```
Comment (derekdowling) [inline index.js]: @hanford this is probably the most controversial decision for you to weigh in on.

**Comment by hanford** [comment]
will do a full review either tomorrow / friday at the latest depending on my work schedule!

**Comment by hanford** [comment]
@derekdowling bunch of stuff came at up @ work on Friday, don't want this PR to lose momentum and will hopefully be able to do a full review either this weekend or early next week.

Appreciate the work / patience! 🙌

File: tests/package.json (hunk 3 above)
Comment (hanford) [reply tests/package.json]: We can keep this change, but it should be this if we go w/ it:

`yarn clean && NODE_ENV=production yarn build && NODE_ENV=production node`

File: tests/package.json (hunk 9 above)
Comment (hanford) [reply tests/package.json]: yes we will! I'm totally open to a lerna / yarn workspace workflow, know of any similar size repo's that are doing it well?

Comment (hanford) [reply tests/package.json]: We should move this conversation to an issue though!

File: tests/package.json (hunk 14 above)
Comment (hanford) [reply tests/package.json]: thanks!

File: index.js (hunk 15 above)
Comment (hanford) [reply index.js]: I think this is fine 🤷‍♂️

**Comment by hanford** [comment]
@derekdowling OK. So, after doing some digging this is working great!

The one thing that I can't seem to get working is `swDest`, and I'm not sure if it's because the above code changes or workbox being in an alpha state.

I'm leaning on the latter, but I want to be certain.

If you head into the now 2.0 example, you'll notice we have this line: https://github.com/hanford/next-offline/blob/master/examples/now2/next.config.js#L7 .. which is where we want the service worker to end up.

If you run `NODE_ENV=production yarn build` you'll see the service worker isn't in `.next/static/service-worker.js` it's instead in `.next/service-woker.js` .. any ideas? 

I can reach out to the workbox crew if we think this is a workbox v5 alpha issue.

Let me know!

**Comment by derekdowling** [comment]
@hanford good catch. As best I can tell from tracing through the Workbox build process, [this](https://github.com/GoogleChrome/workbox/blob/master/packages/workbox-build/src/lib/bundle.js#L30) is causing the `swDest` not to be respected if it contains a folder path prefix.

For example:
```
path.basename('/foo/bar/baz/asdf/quux.html');
// Returns: 'quux.html'

path.basename('/foo/bar/baz/asdf/quux.html', '.html');
// Returns: 'quux'
```

**Comment by derekdowling** [comment]
Filed: https://github.com/GoogleChrome/workbox/issues/2138 to track this. @hanford I don't think this is a blocker on our end if we release this as a v5 Alpha unless we want to try to do some sort of work around.

**Comment by hanford** [comment]
Awesome, thanks @derekdowling. Going to merge this and cut the alpha release tomorrow morning.

**Comment by hanford** [comment]
Thanks again @derekdowling! 👏

Also looks like the next Workbox alpha should include the fix we need for Now 2.0! https://github.com/GoogleChrome/workbox/pull/2140

I've released this as `next-offline@5.0.0-alpha.0`

**Comment by derekdowling** [comment]
Wahoo. Thanks @hanford!

**Comment by jeffposnick** [comment]
Thanks for your feedback and usage of the alpha, all!
